src/main/oauth2/utils.ts

```diff
@@ -0,0 +1,148 @@
+}
+
+export async function fetchAuthorizationServerMetadata(): Promise<{
+    authorizationEndpoint: string;
+    tokenEndpoint: string;
+}> {
+    const response = await fetch(WELL_KNOWN_OAUTH_AUTHORIZATION_SERVER_URL);
+    if (response.status !== 200) {
+        const error = `Failed to fetch OAuth2 authorization server metadata: ${response.status} ${response.statusText}`;
+        log.error(error);
+        throw new Error(error);
+    }
+    const { authorization_endpoint, token_endpoint } = await response.json();
+    if (!authorization_endpoint || !token_endpoint) {
+        const error = "Invalid OAuth2 authorization server metadata";
+        log.error(error);
+        throw new Error(error);
+    }
+    // TODO: Remove this hack once the server is fixed
+    const fixedAuthorizationEndpoint = authorization_endpoint.replaceAll(":8888", "");
+    const fixedTokenEndpoint = token_endpoint.replaceAll(":8888", "");
+    return { authorizationEndpoint: fixedAuthorizationEndpoint, tokenEndpoint: fixedTokenEndpoint };
+}
+
+// Careful with shell.openExternal. https://benjamin-altpeter.de/shell-openexternal-dangers/
+function openInBrowser(url: string) {
+    if (!["https:", "http:"].includes(new URL(url).protocol)) return;
+    // Additional checks to prevent opening arbitrary URLs
+    if (!url.startsWith(OAUTH_AUTHORIZATION_SERVER_URL)) return;
+    shell.openExternal(url);
+}
+
+function createUrlWithQuerystring(baseUrl: string, params: Record<string, string | number | boolean>): string {
+    const queryString = stringify(params);
+    return `${baseUrl}?${queryString}`;
+}
+
+export async function authenticate(): Promise<TokenResponse> {
+    const { authorizationEndpoint, tokenEndpoint } = await fetchAuthorizationServerMetadata();
+    const [code_verifier, code_challenge] = generatePKCE();
+    const redirectHandler = new RedirectHandler();
+    try {
+        const redirect_uri = await redirectHandler.start();
+        const url = createUrlWithQuerystring(authorizationEndpoint, {
+            client_id: OAUTH_CLIENT_ID,
+            response_type: "code",
+            redirect_uri,
+            code_challenge,
+            code_challenge_method: "S256",
+        });
+        openInBrowser(url);
+        const callbackUrl = await redirectHandler.waitForCallback();
+        log.debug(`Received callback URL: ${callbackUrl}`);
+        const code = callbackUrl.searchParams.get("code");
+        log.debug(`Received OAuth2 code: ${code}`);
+        const tokenUrl = createUrlWithQuerystring(tokenEndpoint, {
+            grant_type: "authorization_code",
+            client_id: OAUTH_CLIENT_ID,
+            code,
+            code_verifier,
+            redirect_uri,
+        });
+        const tokenResponse = await fetch(tokenUrl, {
+            method: "POST",
+        });
+        if (tokenResponse.status !== 200) {
+            const error = `Failed to fetch OAuth2 token: ${tokenResponse.status} ${tokenResponse.statusText}`;
```

Comment:
I think it would also be valuable to grab the response's body `await tokenResponse.text()` and log that as well. Because many things can get you a 400, and the body will tell you more, like "code verifier invalid length" or whatnot